### PR TITLE
[RFR] Documentation for the Ldap component

### PR DESCRIPTION
**Pull request by csarrazi**
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes |
| Applies to | 2.8, 3.0 |
| Fixed tickets | #6201 #5756 |

### CODE REVIEW

**Comment by javiereguiluz** [comment]
@csarrazi thanks for this!! Bootstrapping documentation for a new component is very hard ... but at the same time very much appreciated!

I have a global question about the proposed docs. Sometimes you write `LDAP server` but most of the times you write `Ldap server`. Wikipedia [always uses LDAP](https://en.wikipedia.org/wiki/Lightweight_Directory_Access_Protocol) as an acronym. I propose to do the same ... except when referring to the name of the component, which apparently is `Ldap`. Thanks!

File: components/ldap.rst [hunk 1]
```diff
@@ -0,0 +1,66 @@
+.. index::
+   single: Ldap
+   single: Components; Ldap
+
+The Ldap Component
+==================
+
+    The Ldap component provides a means to connect to an Ldap server (OpenLdap or Active Directory).
```
Comment (javiereguiluz) [inline components/ldap.rst]: This description should be very concise, so you did it great. However, if this component adds LDAP support to the Symfony Security component, maybe we should mention it. Right now it looks like this component is stand-alone and not integrated with Symfony security.

**Comment by csarrazi** [comment]
@javiereguiluz Yes, that's what is intended. However, as I wrote this quite late in the evening, I wasn't very consistent. Same thing with `a` / `an` before `LDAP`.

File: components/ldap.rst (hunk 1 above)
Comment (csarrazi) [inline components/ldap.rst]: The component is indeed standalone. The security component actually is the one which integrates LDAP functionality in Symfony.

Comment (javiereguiluz) [inline components/ldap.rst]: :+1:

Comment (csarrazi) [inline components/ldap.rst]: Oh, and I'm actually currently writing the doc for the Security component, too.

**Comment by csarrazi** [comment]
Still needs to pass on all files to normalize `Ldap` <-> `LDAP`, and `a` -> `an`.

Also, `cookbook/security/ldap.rst` still needs some more work (the examples still need to be adapted, I only added stubs for now).

**Comment by iltar** [comment]
Is there anything left to do on this PR? We're in the progress of updating our applications to Symfony 3 and the ldap packages used do not support 3.0 thus we will have to implement our own. Sadly the documentation is still missing :(

**Comment by csarrazi** [comment]
The only remaining part is the cookbook, which is not finalized. Indeed, the configuration sections at the end is not ready yet: Only a few yaml examples are ready, and the configuration for other formats (php, xml) is definitely not ready, as I generally do not use these.

**Comment by csarrazi** [comment]
Done. @javiereguiluz could you take a look at it?

**Comment by csarrazi** [comment]
By the way, this documentation is for version 2.8 and 3.0 of the Ldap component. Changes brought in 3.1 will be addressed in another PR.

File: cookbook/security/ldap.rst [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+
+* The ``http_basic_ldap`` authentication provider, for authenticating
+  against an LDAP server using HTTP Basic. Like all other
+  authentication providers, it can be used with any user provider.
+
+This means that the following scenarios will work:
+
+* Checking a user's password and fetching user information against an
+  LDAP server. This can be done using both the LDAP user provider and
+  either the LDAP form login or LDAP http basic authentication providers.
```
Comment (javiereguiluz) [inline cookbook/security/ldap.rst]: `http` -> `HTTP`

File: cookbook/security/ldap.rst [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+  information from another source (database using FOSUserBundle, for
+  example).
+
+* Loading user information from an LDAP server, while using another
+  authentication strategy (token-based pre-authentication, for example).
+
+Ldap Configuration Reference
+----------------------------
+
+To see the full LDAP (``form_login_ldap``, ``http_basic_ldap``, ``ldap``)
```
Comment (javiereguiluz) [inline cookbook/security/ldap.rst]: I'd reword this:

```
+To see the full LDAP (``form_login_ldap``, ``http_basic_ldap``, ``ldap``)
+configuration reference, see :doc:`/reference/configuration/security`.
```

like this:

```
See :doc:`/reference/configuration/security` for the full LDAP
configuration reference (``form_login_ldap``, ``http_basic_ldap``, ``ldap``).
```

File: cookbook/security/ldap.rst [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+                        uid_key: uid
+
+The ``ldap`` user provider supports many different configuration options:
+
+service
+.......
+
+**type**: ``string`` **default**: ``ldap``
+
+This is the name of your configured LDAP client.
```
Comment (javiereguiluz) [inline cookbook/security/ldap.rst]: I miss here some hints about picking this name. Something like:

```
This is the name of your configured LDAP client. You can freely chose the
name, but it must be unique in your application and it cannot start with a number
or contain white spaces.
```

File: cookbook/security/ldap.rst [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+Authenticating against an LDAP server
+=====================================
+
+Symfony provides different means to work with an LDAP server.
+
+The Security component offers:
+
+* The ``ldap`` user provider, using the
+  :class:`Symfony\\Component\\Security\\Core\\User\\LdapUserProvider``
```
Comment (javiereguiluz) [inline cookbook/security/ldap.rst]: Minor syntax issue: double backticks at the end of this line should be single backtick.

**Comment by javiereguiluz** [comment]
I left some unimportant comments but overall I like this documentation a lot: you went straight to the point, everything is clearly explained and there are a ton of useful examples. You did a truly amazing job here. Thanks!

File: components/ldap.rst [hunk 6]
```diff
@@ -0,0 +1,66 @@
+.. index::
+   single: Ldap
+   single: Components; Ldap
+
+The Ldap Component
+==================
+
+    The Ldap component provides a means to connect to an LDAP server (OpenLDAP or Active Directory).
```
Comment (HeahDude) [inline components/ldap.rst]: It sounds weird to use an "a" with a plural, shouldn't it be "provides the means"?

Comment (csarrazi) [inline components/ldap.rst]: Actually, this is how the word `means` works. In its singular form, it still takes an `s`. :)

Also, as the Ldap component is not the only means to connect to an LDAP server, the article `a` is correct, here. :)

File: cookbook/security/ldap.rst (hunk 3 above)
Comment (csarrazi) [inline cookbook/security/ldap.rst]: Agreed.

File: cookbook/security/ldap.rst (hunk 2 above)
Comment (csarrazi) [inline cookbook/security/ldap.rst]: Done

**Comment by csarrazi** [comment]
Comments corrected.

By the way, the PR for 3.1 will be quite small in comparison, as the only difference will be the description of the new way to define the `ldap` service. However, it first needs this PR to be merged in all branches (2.8, 3.0, master/3.1).

**Comment by csarrazi** [comment]
Are there still missing things here? I'd like to be able to work on adding the doc for version 3.1 of the Ldap component before the 3.1 release.

Ping @javiereguiluz

File: components/ldap.rst [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+Usage
+-----
+
+The :class:`Symfony\\Component\\Ldap\\LdapClient` class provides methods
+to authenticate and query against an LDAP server.
+
+The :class:`Symfony\\Component\\Ldap\\LdapClient` class can be configured
+using the following options:
+
+* host: IP or hostname of the LDAP server
```
Comment (xabbuh) [inline components/ldap.rst]: I would enclose the option names with literals (double backticks).

File: cookbook/security/ldap.rst [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+            ->addArgument(389)
+            ->addArgument(3)
+            ->addArgument(false)
+            ->addArgument(true);
+
+        $container
+            ->register('newsletter_manager', 'NewsletterManager')
+            ->addMethodCall('setMailer', array(new Reference('mailer')));
+
+Fetching users using the LDAP user provider
```
Comment (xabbuh) [inline cookbook/security/ldap.rst]: Fetching Users Using the LDAP User Provider

File: cookbook/security/ldap.rst [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+
+            providers:
+                my_ldap:
+                    ldap:
+                        service: ldap
+                        base_dn: dc=example,dc=com
+                        search_dn: "cn=read-only-admin,dc=example,dc=com"
+                        search_password: password
+                        default_roles: ROLE_USER
+                        uid_key: uid
```
Comment (xabbuh) [inline cookbook/security/ldap.rst]: missing XML and PHP config examples

File: cookbook/security/ldap.rst [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+
+The ``ldap`` user provider supports many different configuration options:
```
Comment (xabbuh) [inline cookbook/security/ldap.rst]: imo the following lines should only be part of the reference

Comment (csarrazi) [inline cookbook/security/ldap.rst]: What do you mean, by that? Move and keep all these into the security reference?

File: cookbook/security/ldap.rst (hunk 9 above)
Comment (csarrazi) [inline cookbook/security/ldap.rst]: Done

File: components/ldap.rst [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+* ``port``: Port used to access the LDAP server
+
+* ``version``: The version of the LDAP protocol to use
+
+* ``useSsl``: Whether or not to secure the connection using SSL
+
+* ``useStartTls``: Whether or not to secure the connection using StartTLS
+
+* ``optReferrals``: Specifies whether to automatically follow referrals
+  returned by the LDAP server
```
Comment (xabbuh) [inline components/ldap.rst]: Sorry, I didn't notice during my last review, but I think having a definition list would better fit what we are trying here:

```rst
``host``
    IP or hostname of the LDAP server

``port``
    Port used to access the LDAP server

``version``
    The version of the LDAP protocol to use

``useSsl``
    Whether or not to secure the connection using SSL

``useStartTls``
    Whether or not to secure the connection using StartTLS

``optReferrals``
    Specifies whether to automatically follow referrals returned by the LDAP server
```

File: cookbook/security/ldap.rst [hunk 12]
```diff
@@ -0,0 +1,384 @@
...
+            ->addArgument(389)
+            ->addArgument(3)
+            ->addArgument(false)
+            ->addArgument(true);
+
+        $container
+            ->register('newsletter_manager', 'NewsletterManager')
+            ->addMethodCall('setMailer', array(new Reference('mailer')));
+
+Fetching users using the LDAP User Provider
```
Comment (xabbuh) [inline cookbook/security/ldap.rst]: Fetching Users Using the LDAP User Provider

File: components/ldap.rst (hunk 11 above)
Comment (csarrazi) [inline components/ldap.rst]: Done.

File: cookbook/security/ldap.rst (hunk 12 above)
Comment (csarrazi) [inline cookbook/security/ldap.rst]: Done.

**Comment by csarrazi** [comment]
Anything else to do @xabbuh, @javiereguiluz?

**Comment by javiereguiluz** [comment]
@csarrazi nothing else for me ... except saying **thank you** for your great work!

I'm sure that doc maintainers will merge this soon. We want to document lots of things before the 3.1 release in 5 days.

**Comment by csarrazi** [comment]
Yeah. And this needs to be merged so I can update the docs for 3.1.

**Comment by xabbuh** [comment]
@javiereguiluz I counted your [last comment](https://github.com/symfony/symfony-docs/pull/6402#issuecomment-221870164) as a +1 vote. So I am going to merge here.

**Comment by xabbuh** [comment]
Thanks for your work on this new feature!

**Comment by csarrazi** [comment]
You're welcome. I'll work on the 3.1 doc asap!

**Comment by xabbuh** [comment]
@csarrazi I merged your changes up into the `3.1`  branch.
